### PR TITLE
Fix module loading/unloading and MPI/MMI session lifetime

### PR DESCRIPTION
**Pull request by robertschaedler3**
## Description

- Fix the `MPI` layer to have a single `ModulesManager` which loads/unloads modules only when MPI is initialized/shutdown by the agent. 
- Create `MpiSession` class to hold the`MmiSession`'s for a client (returned as `MPI_HANDLE` from `MpiOpen()`)
- Create `MmiSession` class to hold the `MMI_HANDLE` for each module

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-osconfig/blob/main/CONTRIBUTING.md).
- [x] I added unit-tests to validate my changes. All unit tests are passing.
- [x] I have merged the latest `main` branch prior to this PR submission.
- [x] I submitted this PR against the `main` branch.

### CODE REVIEW

File: src/platform/modulesmanager/src/ModulesManager.cpp [hunk 1]
```diff
@@ -88,10 +100,16 @@ MPI_HANDLE MpiOpen(
 
 void MpiClose(MPI_HANDLE clientSession)
 {
-    if (mpiMap.end() != mpiMap.find(clientSession))
+    ModulesManager* modulesManager = reinterpret_cast<ModulesManager*>(clientSession);
+
+    if ((nullptr != modulesManager) && (mpiMap.end() != mpiMap.find(modulesManager->GetClientName())))
```
Comment (MariusNi) [inline src/platform/modulesmanager/src/ModulesManager.cpp:105]: Let's imagine this sequence, all for same client:

1. MpiOpen
2. MpiGetReported
3. MpiClose

The 2nd call increments the ref count on the session opened by 1st then decrements it. The 3rd call decrements the ref count once more and when the ref count goes to 0 the MM object for this client session self-destroys itself.

Consider using ref counting and self-release on these objects, COM style?

File: src/modules/ztsi/src/lib/Ztsi.cpp [hunk 2]
```diff
@@ -41,6 +41,7 @@ int SerializeJsonObject(MMI_JSON_STRING* payload, int* payloadSizeBytes, unsigne
     rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
     document.Accept(writer);
 
+    // if ((maxPayloadSizeBytes > 0) && (buffer.GetSize() > maxPayloadSizeBytes))
```
Comment (MariusNi) [inline src/modules/ztsi/src/lib/Ztsi.cpp:44]: > // if ((maxPayloadSizeBytes > 0) && (buffer.GetSize() > maxPayloadSizeBytes)) [](http://example.com/codeflow?start=3&length=78)

Remove, don't comment out code. Re-add later if necessary.

File: src/platform/modulesmanager/inc/ModulesManager.h [hunk 3]
```diff
@@ -49,59 +44,59 @@ class ModulesManagerLog
...
 
 protected:
-    struct ModuleMetadata
-    {
-        std::shared_ptr<ManagementModule> module;
-        std::chrono::system_clock::time_point lastOperation;
-        bool operationInProgress;
-    };
+    // Component name -> vector of reported components according to osconfig.json
+    std::map<std::string, std::unordered_set<std::string>> m_reported;
```
Comment (MariusNi) [inline src/platform/modulesmanager/inc/ModulesManager.h:62]: > m_reported; [](http://example.com/codeflow?start=58&length=12)

Name does not reflect the meaning. How about: m_reportedComponents?

(had to read the comment to understand what this "reported" means)

File: src/platform/modulesmanager/inc/ModulesManager.h [hunk 4]
```diff
@@ -49,59 +44,59 @@ class ModulesManagerLog
...
 
-    // Module mapping componentName <-> ModuleMetadata
-    std::map<std::string, ModuleMetadata> modMap;
-    std::vector<ModuleMetadata> modulesToUnload;
+    // Component name -> module name
+    std::map<std::string, std::string> m_componentToModule;
```
Comment (MariusNi) [inline src/platform/modulesmanager/inc/ModulesManager.h:65]: >m_componentToModule [](http://example.com/codeflow?start=39&length=19)

This is a string, a state. Not a change from one state to another. Change to m_moduleComponentName or similar?

File: src/platform/modulesmanager/inc/ModulesManager.h [hunk 5]
```diff
@@ -49,59 +44,59 @@ class ModulesManagerLog
...
-    static std::string moduleDirectory;
-    static std::string configurationJsonFile;
+    // Module name -> ManagementModule
+    std::map<std::string, std::shared_ptr<ManagementModule>> m_modules;
 
-    std::string clientName;
-    unsigned int maxPayloadSizeBytes;
+    int SetReportedObjects(const std::string& configJson);
+    void ScheduleModuleUnload(std::shared_ptr<ManagementModule> mm);
+    void SetComponentsForModule(const std::string& moduleName, const std::vector<std::string>& components, bool replace = false);
```
Comment (MariusNi) [inline src/platform/modulesmanager/inc/ModulesManager.h:72]: >void SetComponentsForModule [](http://example.com/codeflow?start=4&length=27)

Modules implement component. The MM reads those from each module. The name if this method is misleading.

How about:

ReadModulesComponents

?

File: src/platform/modulesmanager/src/ManagementModule.cpp [hunk 6]
```diff
@@ -7,514 +7,476 @@
...
 {
-    int status = MMI_OK;
-
-    // Validate payload before calling MmiSet
-    if (IsValidMimObjectPayload(payload, payloadSizeBytes, ModulesManagerLog::Get()))
+    if (m_module)
     {
-        LoadModule();
-        status = mmiSet(mmiHandle, componentName, objectName, payload, payloadSizeBytes);
+        m_handle = m_module->CallMmiOpen(m_clientName.c_str(), m_maxPayloadSizeBytes);
```
Comment (MariusNi) [inline src/platform/modulesmanager/src/ManagementModule.cpp:457]: > m_handle = m_module->CallMmiOpen(m_clientName.c_str(), m_maxPayloadSizeBytes); [](http://example.com/codeflow?start=7&length=79)

Dangerous to invoke an external function such as MmiOpen from within constructor.

Leave constructors simple and move such initialization code to some Initialize() member function. Here and elsewhere.

File: src/platform/modulesmanager/inc/ModulesManager.h (hunk 4 above)
Comment (robertschaedler3) [reply src/platform/modulesmanager/inc/ModulesManager.h]: Agreed, and changed to `m_moduleComponentName`

File: src/platform/modulesmanager/src/ManagementModule.cpp [hunk 7]
```diff
@@ -7,514 +7,476 @@
...
+    m_info.userAccount= 0;
 
-    // Get the Management Module info and validate schema
-    Mmi_GetInfo mmiGetInfo = reinterpret_cast<Mmi_GetInfo>(dlsym(handle, MmiFuncMmiGetInfo.c_str()));
-    Mmi_Free mmiFree = reinterpret_cast<Mmi_Free>(dlsym(handle, MmiFuncMmiFree.c_str()));
-    MMI_JSON_STRING retJson = nullptr;
-    int retJsonSize = 0;
-    int retMmiGetInfo = MMI_OK;
-    ScopeGuard sg{[&]()
+    void* m_handle = dlopen(path.c_str(), RTLD_LAZY);
```
Comment (MariusNi) [inline src/platform/modulesmanager/src/ManagementModule.cpp:55]: > void* m_handle = dlopen(path.c_str(), RTLD_LAZY); [](http://example.com/codeflow?start=3&length=50)

Same comment.... Do not make calls that can fail from inside a constructor/destructor.

**Review by MariusNi — APPROVED** [review]
:shipit:

**Pull request merged**